### PR TITLE
fix(aio): activate ServiceWorker updates asap

### LINKED ISSUE #17539: Improve the service worker update behavior
## Current behavior
Whenever we submit a change to angular.io (often multiple-time per day), we generate a new service worker manifest which is deployed along with the new build on the production servers.

The next time user reloads angular.io, the service worker picks up the manifest change, fetches the update, and stages it for activation.

We can't just activate the new version of the app because the current version is already running. Reloading the app at random times would result in a bad user experience, so we are avoiding it. Instead, we just use a [material snack bar](https://material.angular.io/components/component/snack-bar) to show a notification that an update is available.

The pop-up notifications are easy for people to overlook, or are annoying for those that do see them.

## Proposed behavior
- remove the snackbar notification
- add a simple update notification in the footer after the version number. e.g.: Version 4.2.3-build.31733+sha.64fa100 ([update available](url))
- if clicked, ~all active tabs~ the currently active tab with angular.io will be reloaded and the staged update will be activated
- lastly, once an update is staged, we'd start listening on the windown#unload event indicating tab being closed or reloaded, at this time, ~we'd force-reload *all* active angular.io tabs and activate the new update~ we'd activate the staged update.

The proposed behavior is aiming to:
- remove the annoyance that notification cause
- ensure that users use the latest version of the app as soon as possible without sacrificing offline capabilities, caching, and fast startup.
- bring back the refresh-to-update behavior that non-service worker apps have and users expect to work

I should point out that this kind of update strategy is not going to work for every app. Apps that don't have their full state persisted would lose data during the update (e.g. unsubmitted form, expanded menus, scroll position, etc). angular.io doesn't have much user-entered state, so the trade-off is worth it.

### PR DESCRIPTION
**Pull request by gkalpak**
Previouly, whenever a new ServiceWorker update was detected the user was prompted to update (with a notification). This turned out to be more distracting than helpful. Also, one would get notifications on all open browser tabs/windows and had to manually reload each one in order for the whole content (including the app) to be updated.

This commit changes the update strategy as follows:
- Whenever a new update is detected, it is immediately activated (and all tabs/windows will be notified).
- Once an update is activated (regardless of whether the activation was initiated by the current tab/window or not), a flag will be set to do a "full page navigation" the next time the user navigates to a document.

Benefits:
- All tabs/windows are updated asap.
- The updates are applied authomatically, without the user's needing to do anything.
- The updates are applied in a way that:
  a. Ensures that the app and content versions are always compatible.
  b. Does not distract the user from their usual workflow.

NOTE:
The "full page navigation" may cause a flash (while the page is loading from scratch), but this is expected to be minimal, since at that point almost all necessary resources are cached by and served from the ServiceWorker.

Fixes #17539.

### CODE REVIEW

File: aio/src/app/sw-updates/sw-updates.module.ts [hunk 1]
```diff
@@ -1,20 +1,14 @@
 import { NgModule } from '@angular/core';
-import { MdSnackBarModule } from '@angular/material';
```
Comment (petebacondarwin) [inline aio/src/app/sw-updates/sw-updates.module.ts]: Does this removal help lower our bundle sizes or is the snack bar used elsewhere?

File: aio/src/app/sw-updates/sw-updates.service.spec.ts [hunk 2]
```diff
@@ -14,11 +15,13 @@ describe('SwUpdatesService', () => {
 
   // Helpers
   // NOTE:
-  //  Because `SwUpdatesService` uses the `debounceTime` operator, it needs to be instantiated
-  //  inside the `fakeAsync` zone (when `fakeAsync` is used for the test). Thus, we can't run
-  //  `setup()` in a `beforeEach()` block. We use the `run()` helper to call it inside each test' zone.
+  //   Because `SwUpdatesService` uses the `debounceTime` operator, it needs to be instantiated
+  //   inside the `fakeAsync` zone (when `fakeAsync` is used for the test). Thus, we can't run
+  //   `setup()` in a `beforeEach()` block. We use the `run()` helper to call `setup()` inside each
+  //   test's zone.
```
Comment (petebacondarwin) [inline aio/src/app/sw-updates/sw-updates.service.spec.ts]: Why is it the case that we cannot use a `beforeEach`? Surely we can just use the `fakeAsync` helper there too?

```ts
beforeEach(fakeAsync() => {
  injector = ReflectiveInjector.resolveAndCreate([
    { provide: Logger, useClass: MockLogger },
    { provide: NgServiceWorker, useClass: MockNgServiceWorker },
    SwUpdatesService
  ]);
});
```

File: aio/src/app/sw-updates/sw-updates.service.spec.ts [hunk 3]
```diff
@@ -51,163 +54,103 @@ describe('SwUpdatesService', () => {
 
     tick(checkInterval);
     expect(sw.checkForUpdate).toHaveBeenCalled();
+    expect(sw.activateUpdate).not.toHaveBeenCalled();
   })));
 
-  it('should not schedule a new check if there is an update available', fakeAsync(run(() => {
+  it('should activate new updates immediatelly', fakeAsync(run(() => {
```
Comment (petebacondarwin) [inline aio/src/app/sw-updates/sw-updates.service.spec.ts]: typo `immediatelly`

File: aio/src/app/sw-updates/sw-updates.service.ts [hunk 4]
```diff
@@ -17,60 +18,58 @@ import 'rxjs/add/operator/toPromise';
...
- * @method
- * `activateUpdate()` {() => Promise<boolean>} - Activate the latest available update. The returned
- * promise resolves to `true` if an update was activated successfully and `false` if the activation
- * failed (e.g. if there was no update to activate).
+ * `updateActivated` {Observable<string>} - Emit the version hash whenever an update is activated.
  */
 @Injectable()
 export class SwUpdatesService implements OnDestroy {
-  private checkInterval = 1000 * 60 * 60 * 6;   // 6 hours
+  private checkInterval = 1000 /* 60 * 60 */* 6;   // 6 hours
```
Comment (petebacondarwin) [inline aio/src/app/sw-updates/sw-updates.service.ts]: Ahem!

**Comment by petebacondarwin** [comment]
Why use debounce in the `SwUpdatesService` rather than a more simple `setInterval`?

**Comment by petebacondarwin** [comment]
A rebase should fix Travis now we have landed the lighthouse fix

File: aio/src/app/sw-updates/sw-updates.module.ts (hunk 1 above)
Comment (gkalpak) [inline aio/src/app/sw-updates/sw-updates.module.ts]: No, the copy button uses it.

File: aio/src/app/sw-updates/sw-updates.service.spec.ts (hunk 2 above)
Comment (gkalpak) [inline aio/src/app/sw-updates/sw-updates.service.spec.ts]: Not all `it`s use `fakeAsync`.

Comment (petebacondarwin) [inline aio/src/app/sw-updates/sw-updates.service.spec.ts]: Does that matter?

**Comment by gkalpak** [comment]
> Why use debounce in the SwUpdatesService rather than a more simple setInterval?

Because everything is an observable :grin:

File: aio/src/app/sw-updates/sw-updates.service.spec.ts (hunk 2 above)
Comment (gkalpak) [inline aio/src/app/sw-updates/sw-updates.service.spec.ts]: IIRC, if the `beforeEach` is in `fakeAsync` then everything has to be and if everything is in `fakeAsync` then it complains about pending tasks. It is solveable (by flushing the tasks or something), but back when I first wrote the tests it seemed better to do it like this.

(I don't feel strongly about changing it, if you think it is better in an other way.)

**Comment by petebacondarwin** [comment]
I was thinking of something along the lines of

```typescript
@Injectable()
export class SwUpdatesService implements OnDestroy {
  private checkInterval = 1000 * 60 * 60 * 6;   // 6 hours
  private onDestroy = new Subject();
  private checkForUpdateSubj = new Subject();
  updateActivated = this.sw.updates.
                      takeUntil(this.onDestroy)
                      .do(evt => this.log(`Update event: ${JSON.stringify(evt)}`))
                      .filter(evt => evt.type === 'activation')
                      .map(evt => evt.version)
                      .startWith(null);

  constructor(private logger: Logger, private sw: NgServiceWorker) {
    Observable.timer(0, this.checkInterval)
      .takeUntil(this.onDestroy)
      .subscribe(() => this.checkForUpdate());
  }

  ngOnDestroy() {
    this.onDestroy.next();
  }

  private activateUpdate() {
    this.log('Activating update...');
    this.sw.activateUpdate(null).subscribe();
  }

  private checkForUpdate() {
    this.log('Checking for update...');
    this.sw.checkForUpdate()
        // Temp workaround for https://github.com/angular/mobile-toolkit/pull/137.
        // TODO (gkalpak): Remove once #137 is fixed.
        .concat(Observable.of(false)).take(1)
        .do(v => this.log(`Update available: ${v}`))
        .subscribe(v => v && this.activateUpdate());
  }

  private log(message: string) {
    const timestamp = (new Date).toISOString();
    this.logger.log(`[SwUpdates - ${timestamp}]: ${message}`);
  }
}
```

**Review by IgorMinar — APPROVED** [review]

**Comment by IgorMinar** [comment]
@gkalpak this looks good to me, except for the e2e test that seems to be legitimately failing.

**Comment by gkalpak** [comment]
Rebased on master and fixed the failing e2e test.

##
@petebacondarwin, I refactored `updateActivated` as you suggested, as it looks cleaner. I didn't switch to `Observable.timer`, because (a) it doesn't simplify things a lot (and I would have to change several tests) and (b) the current implementation is technically more correct (since we wait for the previous action to complete before initiating a new check). It won't change anything with our current 6-hour interval, but theoretically it is better 😛 (and it did make a difference when I was testing with 6 second intervals).

##
@IgorMinar, the reason that the e2e was failing is because the `aio-code` component imports and uses `MdSnackBar`, but the `MdSnackBarModule` was not imported. (This didn't fail before, because the `SwUpdates` service imported the `MdSnackBarModule`, which is no longer the case.)

What I find unexpected though is that there was no error when building. Shouldn't `yarn build` complain that it doesn't know where to get `MdSnackBar` from? Is it because `<aio-code>` is used in `entryComponents`?

**Pull request merged**